transport_plugins/jlink/iotile_transport_jlink/jlink.py

```diff
@@ -271,32 +289,70 @@ def _parse_conn_string(self, conn_string):
         self._setup_connection(conn_id, connection_string)
 
         try:
-            self._control_info = await self._jlink_async.verify_control_structure(self._device_info, self._control_info)
+            self._control_info = await self._jlink_async.verify_control_structure(
+                self._device_info, self._control_info)
+
+            await self._jlink_async.change_state_flag(
+                self._control_info, AsyncJLink.CONNECTION_BIT, True)
         except HardwareError as e:
             logger.warning("RPC, streaming and tracing won't work, but device still can be flashed or debugged",
                 exc_info=True)
 
         self._connection_id = conn_id
 
     async def disconnect(self, conn_id):
-        await self._jlink_async.close_jlink()
+        try:
+            for interface in JLinkAdapter.POLLING_INTERFACES:
+                if self.opened_interfaces[interface]:
+                    await self.close_interface(conn_id, interface)
+
+            await self._jlink_async.change_state_flag(
+                self._control_info, AsyncJLink.CONNECTION_BIT, False)
+        except pylink.errors.JLinkReadException:
+            logger.warning("Error disconnecting jlink adapter", exc_info=True)
+
         self._teardown_connection(conn_id)
 
     async def open_interface(self, conn_id, interface):
         self._ensure_connection(conn_id, True)
 
+        if not self._control_info:
+            return
+
         if self.opened_interfaces[interface]:
             return
 
         self.opened_interfaces[interface] = True
 
-        if interface in ["rpc", "tracing", "streaming"]:
+        if interface == "tracing":
+            await self._jlink_async.change_state_flag(
+                self._control_info, AsyncJLink.TRACE_BIT, True)
+        elif interface == "streaming":
+            await self._jlink_async.change_state_flag(
+                self._control_info, AsyncJLink.STREAM_BIT, True)
+
+        if interface in JLinkAdapter.POLLING_INTERFACES:
             await self._jlink_async.start_polling(self._control_info)
 
     async def close_interface(self, conn_id, interface):
         self._ensure_connection(conn_id, True)
 
-        if interface in ["rpc", "tracing", "streaming"]:
+        if not self._control_info:
```

Comment:
Same comment as above, this error chomping should not happen here and instead we need to fix the code that is calling `close_interface` gratuitously

Reply:
Yes, you are right